File: src/test/java/uk/gov/justice/hmpps/prison/web/config/AuthAwareAuthenticationConverterTest.java
package uk.gov.justice.hmpps.prison.web.config;

import org.junit.jupiter.api.Test;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static uk.gov.justice.hmpps.prison.security.AuthSource.AUTH;
import static uk.gov.justice.hmpps.prison.security.AuthSource.NONE;

public class AuthAwareAuthenticationConverterTest {

    private AuthAwareAuthenticationConverter authenticationConverter = new AuthAwareAuthenticationConverter();
    private Jwt jwt = mock(Jwt.class);

    @Test
    public void convert_basicUserAttributes_attributesCopied() {
        when(jwt.getClaims()).thenReturn(claims("some_user", "auth", "ROLE_some", null));

        final var authToken = authenticationConverter.convert(jwt);

        assertThat(authToken.getPrincipal()).isEqualTo("some_user");
        assertThat(authToken.getAuthSource()).isEqualTo(AUTH);
        assertThat(authToken.getAuthorities()).containsExactlyInAnyOrder(new SimpleGrantedAuthority("ROLE_some"));
        assertThat(authToken.isClientOnly()).isFalse();
    }

    @Test
    public void convert_basicClientAttributes_attributesCopied() {
        when(jwt.getClaims()).thenReturn(claims(null, "auth", "ROLE_some", "client-details"));
        when(jwt.getSubject()).thenReturn("client-details");

        final var authToken = authenticationConverter.convert(jwt);

        assertThat(authToken.getPrincipal()).isEqualTo(null);
        assertThat(authToken.getAuthSource()).isEqualTo(AUTH);
        assertThat(authToken.getAuthorities()).containsExactlyInAnyOrder(new SimpleGrantedAuthority("ROLE_some"));
        assertThat(authToken.isClientOnly()).isTrue();
    }

    @Test
    public void convert_missingUserName_principalIsNull() {
        when(jwt.getClaims()).thenReturn(claims(null, "some_auth_source", "ROLE_some", null));

        final var authToken = authenticationConverter.convert(jwt);

        assertThat(authToken.getPrincipal()).isEqualTo(null);
    }

    @Test
    public void convert_missingAuthSource_authSourceIsNone() {
        when(jwt.getClaims()).thenReturn(claims("some_user", null, "ROLE_some", null));

        final var authToken = authenticationConverter.convert(jwt);

        assertThat(authToken.getAuthSource()).isEqualTo(NONE);
    }

    @Test
    public void convert_missingAuthorities_noGrantedAuthority() {
        when(jwt.getClaims()).thenReturn(claims("some_user", "some_auth_source", null, null));

        final var authToken = authenticationConverter.convert(jwt);

        assertThat(authToken.getAuthorities()).isEmpty();
    }

    private Map<String, Object> claims(String username, String authSource, String scope, String clientId) {
        Map<String, Object> claims = new HashMap<>();
        if (username != null && !username.isBlank()) {
            claims.put("user_name", username);
        }
        if (authSource != null && !authSource.isBlank()) {
            claims.put("auth_source", authSource);
        }
        if (scope != null && !scope.isBlank()) {
            claims.put("authorities", Set.of(scope));
        }
        if (clientId != null && !clientId.isBlank()) {
            claims.put("client_id", clientId);
        }
        return claims;
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/web/config/AuthAwareAuthenticationConverterTest.kt
package uk.gov.justice.hmpps.prison.web.config

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.kotlin.whenever
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import uk.gov.justice.hmpps.prison.security.AuthSource

class AuthAwareAuthenticationConverterTest {
  private val authenticationConverter = AuthAwareAuthenticationConverter()
  private val jwt: Jwt = mock()

  @Test
  fun convert_basicUserAttributes_attributesCopied() {
    whenever(jwt.claims).thenReturn(claims("some_user", "auth", "ROLE_some", null))
    val authToken = authenticationConverter.convert(jwt)
    assertThat(authToken!!.principal).isEqualTo("some_user")
    assertThat(authToken.authSource).isEqualTo(AuthSource.AUTH)
    assertThat(authToken.authorities).containsExactlyInAnyOrder(SimpleGrantedAuthority("ROLE_some"))
    assertThat(authToken.isClientOnly).isFalse()
  }

  @Test
  fun convert_basicClientAttributes_attributesCopied() {
    whenever(jwt.claims).thenReturn(claims(null, "auth", "ROLE_some", "client-details"))
    whenever(jwt.subject).thenReturn("client-details")
    val authToken = authenticationConverter.convert(jwt)
    assertThat(authToken!!.principal).isEqualTo(null)
    assertThat(authToken.authSource).isEqualTo(AuthSource.AUTH)
    assertThat(authToken.authorities).containsExactlyInAnyOrder(SimpleGrantedAuthority("ROLE_some"))
    assertThat(authToken.isClientOnly).isTrue()
  }

  @Test
  fun convert_missingUserName_principalIsNull() {
    whenever(jwt.claims).thenReturn(claims(null, "some_auth_source", "ROLE_some", null))
    val authToken = authenticationConverter.convert(jwt)
    assertThat(authToken!!.principal).isEqualTo(null)
    assertThat(authToken.authSource).isEqualTo(AuthSource.NONE)
    assertThat(authToken.authorities).containsExactlyInAnyOrder(SimpleGrantedAuthority("ROLE_some"))
    assertThat(authToken.isClientOnly).isFalse()
  }

  @Test
  fun convert_missingAuthSource_authSourceIsNone() {
    whenever(jwt.claims).thenReturn(claims("some_user", null, "ROLE_some", null))
    val authToken = authenticationConverter.convert(jwt)
    assertThat(authToken!!.authSource).isEqualTo(AuthSource.NONE)
  }

  @Test
  fun convert_missingAuthorities_noGrantedAuthority() {
    whenever(jwt.claims).thenReturn(claims("some_user", "some_auth_source", null, null))
    val authToken = authenticationConverter.convert(jwt)
    assertThat(authToken!!.authorities).isEmpty()
  }

  private fun claims(username: String?, authSource: String?, scope: String?, clientId: String?): Map<String, Any> {
    val claims: MutableMap<String, Any> = HashMap()
    if (!username.isNullOrBlank()) {
      claims["user_name"] = username
    }
    if (!authSource.isNullOrBlank()) {
      claims["auth_source"] = authSource
    }
    if (!scope.isNullOrBlank()) {
      claims["authorities"] = setOf(scope)
    }
    if (!clientId.isNullOrBlank()) {
      claims["client_id"] = clientId
    }
    return claims
  }
}
